File: backend/app/repositories/user_repo.py
from sqlalchemy.orm import Session
from datetime import datetime, date
from typing import List, Optional
from app.models.daily_streak import DailyStreak
from app.models.leaderboard import Leaderboard
from app.models.user_attempt import UserAttempt
from app.models.progress import Progress
from app.models.bookmark import Bookmark
from app.models.favorite import Favorite
from app.models.badge import Badge
from app.models.achievement import Achievement

class UserRepository:
    def __init__(self, db: Session):
        self.db = db

    # Streak Operations
    def get_or_create_streak(self, user_id: int) -> DailyStreak:
        streak = self.db.query(DailyStreak).filter(DailyStreak.user_id == user_id).first()
        if not streak:
            streak = DailyStreak(user_id=user_id, current_streak=0, longest_streak=0)
            self.db.add(streak)
            self.db.commit()
            self.db.refresh(streak)
        return streak

    def update_streak(self, user_id: int) -> DailyStreak:
        streak = self.get_or_create_streak(user_id)
        today = date.today()
        if streak.last_activity_date == today:
            return streak # Already updated today
            
        if streak.last_activity_date:
            delta = (today - streak.last_activity_date).days
            if delta == 1:
                streak.current_streak += 1
            elif delta > 1:
                streak.current_streak = 1 # Streak broken, reset to 1
        else:
            streak.current_streak = 1
            
        if streak.current_streak > streak.longest_streak:
            streak.longest_streak = streak.current_streak
            
        streak.last_activity_date = today
        self.db.commit()
        self.db.refresh(streak)
        return streak

    # Leaderboard / XP Operations
    def get_or_create_profile(self, user_id: int, username: str = "Anonymous Student") -> Leaderboard:
        profile = self.db.query(Leaderboard).filter(Leaderboard.user_id == user_id).first()
        if not profile:
            profile = Leaderboard(user_id=user_id, username=username, xp=0, rank=0)
            self.db.add(profile)
            self.db.commit()
            self.db.refresh(profile)
        return profile

    def add_xp(self, user_id: int, xp_amount: int) -> Leaderboard:
        profile = self.get_or_create_profile(user_id)
        profile.xp += xp_amount
        self.db.commit()
        self.db.refresh(profile)
        self.recalculate_ranks()
        return profile

    def recalculate_ranks(self) -> None:
        # Simple dense rank update
        profiles = self.db.query(Leaderboard).order_by(Leaderboard.xp.desc()).all()
        for rank_idx, profile in enumerate(profiles, start=1):
            profile.rank = rank_idx
        self.db.commit()

    def get_leaderboard(self, limit: int = 50) -> List[Leaderboard]:
        return self.db.query(Leaderboard).order_by(Leaderboard.rank.asc()).limit(limit).all()

    # User Attempts
    def record_attempt(self, user_id: int, question_id: int, language_id: int, status: str, runtime: float, memory: int, submitted_code: str) -> UserAttempt:
        # Calculate attempt number
        attempt_count = self.db.query(UserAttempt).filter_by(user_id=user_id, question_id=question_id).count()
        attempt_number = attempt_count + 1
        
        attempt = UserAttempt(
            user_id=user_id,
            question_id=question_id,
            language_id=language_id,
            status=status,
            runtime=runtime,
            memory=memory,
            attempt_number=attempt_number,
            submitted_code=submitted_code
        )
        self.db.add(attempt)
        self.db.commit()
        self.db.refresh(attempt)
        return attempt

    def get_attempts(self, user_id: int, question_id: int) -> List[UserAttempt]:
        return self.db.query(UserAttempt).filter_by(user_id=user_id, question_id=question_id).order_by(UserAttempt.attempt_number.asc()).all()

    def get_attempt_count(self, user_id: int, question_id: int) -> int:
        return self.db.query(UserAttempt).filter_by(user_id=user_id, question_id=question_id).count()

    # Progress Tracking
    def mark_progress(self, user_id: int, question_id: int, status: str) -> Progress:
        prog = self.db.query(Progress).filter_by(user_id=user_id, question_id=question_id).first()
        if not prog:
            prog = Progress(user_id=user_id, question_id=question_id, status=status)
            self.db.add(prog)
        else:
            # Once solved, cannot go back to attempted
            if prog.status != "solved" or status == "solved":
                prog.status = status
        self.db.commit()
        self.db.refresh(prog)
        return prog

    # Bookmarks
    def toggle_bookmark(self, user_id: int, question_id: int) -> bool:
        bookmark = self.db.query(Bookmark).filter_by(user_id=user_id, question_id=question_id).first()
        if bookmark:
            self.db.delete(bookmark)
            self.db.commit()
            return False # Removed
        else:
            bookmark = Bookmark(user_id=user_id, question_id=question_id)
            self.db.add(bookmark)
            self.db.commit()
            return True # Added

    # Favorites
    def toggle_favorite(self, user_id: int, question_id: int) -> bool:
        favorite = self.db.query(Favorite).filter_by(user_id=user_id, question_id=question_id).first()
        if favorite:
            self.db.delete(favorite)
            self.db.commit()
            return False # Removed
        else:
            favorite = Favorite(user_id=user_id, question_id=question_id)
            self.db.add(favorite)
            self.db.commit()
            return True # Added

    # Achievements / Badges
    def award_badge_if_earned(self, user_id: int, badge_name: str) -> Optional[Achievement]:
        badge = self.db.query(Badge).filter(Badge.name == badge_name).first()
        if not badge:
            return None
            
        ach = self.db.query(Achievement).filter_by(user_id=user_id, badge_id=badge.id).first()
        if not ach:
            ach = Achievement(user_id=user_id, badge_id=badge.id)
            self.db.add(ach)
            self.db.commit()
            self.db.refresh(ach)
            return ach
        return None
        
    def get_user_achievements(self, user_id: int) -> List[Badge]:
        return self.db.query(Badge).join(Achievement).filter(Achievement.user_id == user_id).all()

File: backend/app/repositories/user_repo.js
const db = require('../database/session');

class UserRepository {
  // Streak Operations
  static async getOrCreateStreak(userId) {
    const col = await db.collection('daily_streaks');
    const streak = await col.findOne({ user_id: userId });
    if (streak) {
      return streak;
    }
    const id = await db.getNextSequenceValue('streak_id');
    const doc = {
      _id: id,
      id: id,
      user_id: userId,
      current_streak: 0,
      longest_streak: 0,
      last_activity_date: null
    };
    await col.insertOne(doc);
    return doc;
  }

  static async updateStreak(userId) {
    const col = await db.collection('daily_streaks');
    const streak = await this.getOrCreateStreak(userId);
    const todayStr = new Date().toISOString().split('T')[0];

    if (streak.last_activity_date) {
      const lastActStr = new Date(streak.last_activity_date).toISOString().split('T')[0];
      if (lastActStr === todayStr) {
        return streak;
      }

      const lastDate = new Date(lastActStr);
      const todayDate = new Date(todayStr);
      const diffTime = Math.abs(todayDate - lastDate);
      const diffDays = Math.round(diffTime / (1000 * 60 * 60 * 24));

      let currentStreak = streak.current_streak;
      if (diffDays === 1) {
        currentStreak += 1;
      } else {
        currentStreak = 1;
      }

      let longestStreak = streak.longest_streak;
      if (currentStreak > longestStreak) {
        longestStreak = currentStreak;
      }

      await col.updateOne(
        { user_id: userId },
        { $set: { current_streak: currentStreak, longest_streak: longestStreak, last_activity_date: todayStr } }
      );
      return await col.findOne({ user_id: userId });
    } else {
      await col.updateOne(
        { user_id: userId },
        { $set: { current_streak: 1, longest_streak: 1, last_activity_date: todayStr } }
      );
      return await col.findOne({ user_id: userId });
    }
  }

  // Leaderboard / XP Operations
  static async getOrCreateProfile(userId, username = "Anonymous Student") {
    const col = await db.collection('leaderboard');
    const profile = await col.findOne({ user_id: userId });
    if (profile) {
      return profile;
    }
    const id = await db.getNextSequenceValue('leaderboard_id');
    const doc = {
      _id: id,
      id: id,
      user_id: userId,
      username: username,
      xp: 0,
      rank: 0
    };
    await col.insertOne(doc);
    return doc;
  }

  static async addXp(userId, xpAmount) {
    const col = await db.collection('leaderboard');
    const profile = await this.getOrCreateProfile(userId);
    const newXp = profile.xp + xpAmount;
    await col.updateOne({ user_id: userId }, { $set: { xp: newXp } });
    await this.recalculateRanks();
    return await this.getOrCreateProfile(userId);
  }

  static async recalculateRanks() {
    const col = await db.collection('leaderboard');
    const all = await col.find({}).sort({ xp: -1 }).toArray();
    for (let i = 0; i < all.length; i++) {
      const rank = i + 1;
      await col.updateOne({ _id: all[i]._id }, { $set: { rank: rank } });
    }
  }

  static async getLeaderboard(limit = 50) {
    const col = await db.collection('leaderboard');
    const list = await col.find({}).sort({ rank: 1 }).limit(limit).toArray();
    return list.map(row => {
      return {
        user_id: row.user_id,
        username: row.username,
        xp: row.xp,
        rank: row.rank
      };
    });
  }

  // User Attempts
  static async recordAttempt(userId, questionId, languageId, status, runtime, memory, submittedCode) {
    const col = await db.collection('user_attempts');
    const attemptNumber = (await col.countDocuments({ user_id: userId, question_id: questionId })) + 1;
    const id = await db.getNextSequenceValue('attempt_id');
    const doc = {
      _id: id,
      id: id,
      user_id: userId,
      question_id: questionId,
      language_id: languageId,
      status: status,
      runtime: runtime ?? null,
      memory: memory ?? null,
      attempt_number: attemptNumber,
      submitted_code: submittedCode,
      created_at: new Date()
    };
    await col.insertOne(doc);
    return doc;
  }

  static async getAttempts(userId, questionId) {
    const col = await db.collection('user_attempts');
    return await col.find({ user_id: userId, question_id: questionId }).sort({ attempt_number: 1 }).toArray();
  }

  static async getAttemptCount(userId, questionId) {
    const col = await db.collection('user_attempts');
    return await col.countDocuments({ user_id: userId, question_id: questionId });
  }

  // Progress Tracking
  static async markProgress(userId, questionId, status) {
    const col = await db.collection('progress');
    const progress = await col.findOne({ user_id: userId, question_id: questionId });
    if (!progress) {
      const id = await db.getNextSequenceValue('progress_id');
      const doc = {
        _id: id,
        id: id,
        user_id: userId,
        question_id: questionId,
        status: status,
        updated_at: new Date()
      };
      await col.insertOne(doc);
      return doc;
    } else {
      const currentStatus = progress.status;
      if (currentStatus !== "solved" || status === "solved") {
        await col.updateOne(
          { user_id: userId, question_id: questionId },
          { $set: { status: status, updated_at: new Date() } }
        );
        return await col.findOne({ user_id: userId, question_id: questionId });
      }
      return progress;
    }
  }

  // Bookmarks
  static async toggleBookmark(userId, questionId) {
    const col = await db.collection('bookmarks');
    const existing = await col.findOne({ user_id: userId, question_id: questionId });
    if (existing) {
      await col.deleteOne({ user_id: userId, question_id: questionId });
      return false;
    } else {
      const id = await db.getNextSequenceValue('bookmark_id');
      await col.insertOne({ _id: id, id, user_id: userId, question_id: questionId });
      return true;
    }
  }

  // Favorites
  static async toggleFavorite(userId, questionId) {
    const col = await db.collection('favorites');
    const existing = await col.findOne({ user_id: userId, question_id: questionId });
    if (existing) {
      await col.deleteOne({ user_id: userId, question_id: questionId });
      return false;
    } else {
      const id = await db.getNextSequenceValue('favorite_id');
      await col.insertOne({ _id: id, id, user_id: userId, question_id: questionId });
      return true;
    }
  }

  // Achievements / Badges
  static async awardBadgeIfEarned(userId, badgeName) {
    const badgeCol = await db.collection('badges');
    const badge = await badgeCol.findOne({ name: badgeName });
    if (!badge) {
      return null;
    }
    const badgeId = badge._id;

    const achCol = await db.collection('achievements');
    const ach = await achCol.findOne({ user_id: userId, badge_id: badgeId });
    if (!ach) {
      const id = await db.getNextSequenceValue('achievement_id');
      const doc = {
        _id: id,
        id: id,
        user_id: userId,
        badge_id: badgeId,
        earned_at: new Date()
      };
      await achCol.insertOne(doc);
      return doc;
    }
    return null;
  }

  static async getUserAchievements(userId) {
    const achCol = await db.collection('achievements');
    const achievements = await achCol.find({ user_id: userId }).toArray();
    if (achievements.length === 0) {
      return [];
    }
    const badgeIds = achievements.map(a => a.badge_id);
    const badgeCol = await db.collection('badges');
    return await badgeCol.find({ _id: { $in: badgeIds } }).toArray();
  }
}

module.exports = UserRepository;
